src/phyltr/commands/annotate.py

```diff
@@ -1,73 +1,55 @@
-        from the first tree in the treestream.
-
-    files
-        A whitespace-separated list of filenames to read treestreams from.
-        Use a filename of "-" to read from stdin.  If no filenames are
-        specified, the treestream will be read from stdin.
-"""
-
 import csv
 import sys
 
 from phyltr.commands.base import PhyltrCommand
 from phyltr.plumbing.sinks import NullSink
-from phyltr.utils.phyltroptparse import OptionParser
+from phyltr.utils.misc import dicts_from_csv
 
-class Annotate(PhyltrCommand):
-
-    parser = OptionParser(__doc__, prog="phyltr annotate")
-    parser.add_option('-e', '--extract', default=False, action="store_true", help="Extract data from annotated tree to file.")
-    parser.add_option('-f', '--file', dest="filename", help="File to read/write annotation data from/to.")
-    parser.add_option('-k', '--key', dest="key", help="Name of column in annotation file to match against taxon names")
-    parser.add_option('-m', '--multiple', default=False, action="store_true")
-
-    def __init__(self, filename, key=None, extract=False, multiple=False):
-        self.filename = filename
-        self.key = key
-        self.extract = extract
-        self.multiple = multiple
 
+class Annotate(PhyltrCommand):
+    """Annotate a the trees in a tree stream with information from a file"""
+    __options__ = [
+        (
+            ('-e', '--extract'),
+            dict(
+                default=False,
+                action="store_true",
+                help="Build a CSV file of information from a treestream, i.e. reverse the"
+                     "standard behaviour")),
+        (
+            ('-f', '--file'),
+            dict(dest="filename", help="File to read/write annotation data from/to.")),
+        (
+            ('-k', '--key'),
+            dict(help="Fieldname which corresponds to tree taxon names, used to link lines of the "
+                      "csv file to tree nodes.")),
+        (
+            ('-m', '--multiple'),
+            dict(
+                default=False, action="store_true",
+                help="If set, when --extract is used, information from each tree in the treestream "
+                     "will be added to the file, with a `tree_number` column used to disambiguate. "
+                     "When not set, information is extracted only from the first tree in the "
+                     "treestream.")),
+    ]
+
+    def __init__(self, **kw):
+        PhyltrCommand.__init__(self, **kw)
         self.n = 0
+        self.annotations = {}
 
-        if not self.extract:
-            self.read_annotation_file()
-
-    @classmethod 
-    def init_from_opts(cls, options, files):
-        annotate = cls(options.filename, options.key, options.extract, options.multiple)
-        if annotate.extract and (annotate.filename == "-" or not annotate.filename):
+        if self.opts.extract and (self.opts.filename == "-" or not self.opts.filename):
```

Comment:
There's almost no need now to overwrite `init_from_opts`, because the complete `argparse.Namespace` with all options is available in the class as `opts` attribute.